rules/do-you-use-these-useful-react-hooks/rule.md

```diff
@@ -1,161 +1,259 @@
+- **Server Rendering**: `useEffect` is client-centric and doesn't engage during server-side rendering
+
+Read more about `useEffect` on the [offical docs](https://react.dev/reference/react/useEffect)
+
+## 3. **useContext**: Using Context Seamlessly 🌍
+
+`useContext` is a pivotal React Hook, giving you the power to both read and subscribe to context values right within your component. Below is a breakdown based on the details you provided:
+
+```jsx
+import { createContext, useContext } from 'react';
+
+// Create a context
+const ThemeContext = createContext({
+  background: 'light',
+  foreground: 'dark',
+});
+
+function ThemedButton() {
+  const theme = useContext(ThemeContext);
+  return (
+    <button style={{ background: theme.background, color: theme.foreground }}>
+      I am styled by theme context!
+    </button>
+  );
+}
+
+export default function App() {
+  return (
+    <ThemeContext.Provider value={{ background: 'black', foreground: 'white' }}>
+      <ThemedButton />
+    </ThemeContext.Provider>
+  );
+}
+```
+**Figure: A Themed button example using useContext**
+
+
+### ✅ Recommended Usage
+
+- **Reading and Subscribing to Context**: Directly access and subscribe to context values straight from your component
+- **Passing Data Deeply**: Bypass manual prop-drilling, letting you transmit data deeply through the component hierarchy
+- **Updating Data Passed via Context**: Easily modify context values and integrate them with state for seamless updates across various components
+- **Specifying a Fallback Default Value**: If no context provider is present upstream, `useContext` will resort to the default value established during context creation
+- **Overriding Context**: For tailored requirements, override the context in specific parts of the component tree by enveloping it in a provider with a distinct value
+- **Optimizing Re-renders**: Amplify performance when transferring objects or functions through context using techniques such as `useCallback` and `useMemo`
+
+### ⚠️ Pitfalls
+
+- **Provider Position**: The context search strategy employed by `useContext` is top-down, always eyeing the nearest provider. It disregards providers present in the invoking component
+- **Re-rendering Children**: Context shifts compel React to re-render all child components stemming from the provider with a changed value. The assessment hinges on the `Object.is` comparison, meaning that even `memo` cannot fend off updates stemming from refreshed context values
+- **Duplicate Modules**: Be wary of build systems churning out duplicate modules (e.g., due to symlinks). This can disintegrate context as both the provider and consumer must be the exact same object, passing the `===` comparison test
+- **Provider Without a Value**: An absent value prop in a provider translates to `value={undefined}`. The default from `createContext(defaultValue)` comes into play only when there's a complete absence of a matching provider
```

Comment:
```suggestion
- **Provider Without a Value**: An absent value prop in a provider translates to `value={undefined}`. The default from `createContext(defaultValue)` comes into play only when there's a complete absence of a matching provider
- **Provider Cannot be Accessed**: If `useContext` is used in a component that is not wrapped by a provider, this can cause client-side errors as the value accessed will be null
```